Fix: Enhance ModelConfig for kv cache size calculations

qixiang-99 opened the pull request:
# [[nvbug](https://nvbugspro.nvidia.com/bug/5380887)][fix] \<Enhance ModelConfig for kv cache size calculations\>


## Description

- Added support for setting num_kv_heads in ModelConfig.
- Implemented logic to determine head_size from head_size or head_dim in pretrained_config.
- Added warnings for default values when configurations are not set.
- Temporarily remove a Gemma3 1b chunk prefill test case because the kernel compatibility issue.
## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added detailed statistics for free cache blocks per attention window size, now accessible in both C++ and Python interfaces.

* **Improvements**
  * Enhanced model configuration handling for cache size calculations, providing more robust and explicit parameter management.
  * Improved resource management logic to better support Variable Sliding Window Attention (VSWA) by accurately tracking and reporting free blocks and available tokens.

* **Tests**
  * Added and updated integration tests, including new test cases for chunked prefill scenarios (currently skipped due to an upstream issue).

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

qixiang-99 committed:
Fix: Enhance ModelConfig for kv cache size calculations

- Added support for setting num_kv_heads in ModelConfig.
- Implemented logic to determine head_size from head_size or head_dim in pretrained_config.
- Added warnings for default values when configurations are not set.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
Remove chunk prefill test case due to kernel incompatibility.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
Remove accidentally added skip.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #11402](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11402/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8433](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8433/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/8433/test-results/rerun_report.html) for details.

qixiang-99 commented on tests/integration/defs/accuracy/test_llm_api_pytorch.py:
```diff
@@ -72,7 +72,6 @@ def test_nvfp4_streaming(self, stream_interval):
             task.evaluate(llm, streaming=True)
 
 
-@skip_post_blackwell  # TODO: remove this skip after this nvbug is fixed: https://nvbugspro.nvidia.com/bug/5295470
```
Remove the accidentally added skip in [PR5151 ](https://github.com/NVIDIA/TensorRT-LLM/pull/5151)

qixiang-99 commented:
> ```
> Temporarily remove a Gemma3 1b chunk prefill test case because the kernel compatibility issue.
> ```
> 
> Do colleagues working on kernels know about this? Should we keep the test around and waive it with a bug instead of removing? Chunked prefill is important and we'll eventually needed.

I will add it back and waive it with related nvbug link, thanks for the suggestion!

brb-nv reviewed (approved):
LGTM

qixiang-99 committed:
Skip chunk prefill test case due to kernel incompatibility.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #11466](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11466/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8485](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8485/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #11471](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11471/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #11678](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11678/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11678](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11678/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8647](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8647/) completed with status: 'FAILURE'

qixiang-99 committed:
fix:
- expose `num_free_blocks_per_window_size` via kv_cache_stats.
- with `num_free_blocks_per_window_size` , update `get_num_free_blocks()` and `get_num_available_tokens()`

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #11856](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11856/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #11881](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11881/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #11963](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11963/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12021](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12021/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8927](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8927/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/8927/test-results/rerun_report.html) for details.

tensorrt-cicd commented:
[PR_Github #12113](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12113/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12113](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12113/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #12021](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/12021/) for commit `b9855b9`
